File: StatusTrackerFunction/src/main/java/com/sample/service/StatusTrackerService.java
package com.sample.service;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import com.sample.model.Status;
import com.sample.model.StatusHistoryItem;

import software.amazon.awssdk.services.comprehend.ComprehendClient;
import software.amazon.awssdk.services.comprehend.model.DetectSentimentRequest;
import software.amazon.awssdk.services.comprehend.model.DetectSentimentResponse;
import software.amazon.awssdk.services.comprehend.model.SentimentType;

/**
 * Service for managing client status tracking in financial services.
 * This service handles operations related to client status records including
 * creation, retrieval, updates, and search functionality.
 */
public class StatusTrackerService {
    
    // Status types for financial services
    public static final String STATUS_TYPE_ACCOUNT_OPENING = "account_opening";
    public static final String STATUS_TYPE_PORTFOLIO_REVIEW = "portfolio_review";
    public static final String STATUS_TYPE_FINANCIAL_PLAN = "financial_plan";
    public static final String STATUS_TYPE_TRADE_EXECUTION = "trade_execution";
    public static final String STATUS_TYPE_FUND_TRANSFER = "fund_transfer";
    public static final String STATUS_TYPE_TAX_DOCUMENT = "tax_document";
    public static final String STATUS_TYPE_COMPLIANCE_CHECK = "compliance_check";
    public static final String STATUS_TYPE_CLIENT_MEETING = "client_meeting";
    
    // Status stages
    public static final String STAGE_INITIATED = "initiated";
    public static final String STAGE_IN_PROGRESS = "in_progress";
    public static final String STAGE_PENDING_REVIEW = "pending_review";
    public static final String STAGE_PENDING_CLIENT_ACTION = "pending_client_action";
    public static final String STAGE_PENDING_ADVISOR_ACTION = "pending_advisor_action";
    public static final String STAGE_PENDING_THIRD_PARTY = "pending_third_party";
    public static final String STAGE_COMPLETED = "completed";
    public static final String STAGE_CANCELLED = "cancelled";
    public static final String STAGE_ON_HOLD = "on_hold";
    
    // AWS clients
    private final ComprehendClient comprehendClient;
    private final StatusDynamoDBService dynamoDBService;
    private final boolean isTestMode;
    
    /**
     * Default constructor that initializes AWS clients.
     */
    public StatusTrackerService() {
        // Initialize AWS clients
        this.comprehendClient = ComprehendClient.builder().build();
        this.dynamoDBService = new StatusDynamoDBService();
        this.isTestMode = false;
    }
    
    /**
     * Constructor with dependency injection for testing.
     * 
     * @param dynamoDBService The DynamoDB service to use
     */
    public StatusTrackerService(StatusDynamoDBService dynamoDBService) {
        this.dynamoDBService = dynamoDBService;
        // In test mode, we don't need a real Comprehend client
        this.comprehendClient = null;
        this.isTestMode = true;
    }
    
    /**
     * Creates a new status entry from the provided data.
     */
    public Status createStatus(Map<String, Object> statusData) {
        try {
            // Create a new status object
            Status status = new Status();
            
            // Generate a unique ID
            String statusId = UUID.randomUUID().toString();
            status.setStatusId(statusId);
            
            // Set core fields from the provided data
            status.setClientId((String) statusData.get("clientId"));
            status.setAdvisorId((String) statusData.get("advisorId"));
            status.setStatusType((String) statusData.get("statusType"));
            status.setCurrentStage(STAGE_INITIATED);
            status.setStatusSummary((String) statusData.get("statusSummary"));
            
            // Set timestamps
            Instant now = Instant.now();
            status.setCreatedDate(now.toString());
            status.setLastUpdatedDate(now.toString());
            
            // Set user info
            status.setCreatedBy((String) statusData.get("createdBy"));
            status.setLastUpdatedBy((String) statusData.get("createdBy"));
            
            // Set source ID if provided
            if (statusData.containsKey("sourceId")) {
                status.setSourceId((String) statusData.get("sourceId"));
            }
            
            // Set tracking ID if provided, otherwise use the generated one
            if (statusData.containsKey("trackingId")) {
                status.setTrackingId((String) statusData.get("trackingId"));
            }
            
            // Set optional fields if provided
            if (statusData.containsKey("statusDetails")) {
                status.setStatusDetails((Map<String, Object>) statusData.get("statusDetails"));
            } else {
                status.setStatusDetails(new HashMap<>());
            }
            
            if (statusData.containsKey("requiredActions")) {
                status.setRequiredActions((List<String>) statusData.get("requiredActions"));
            } else {
                status.setRequiredActions(new ArrayList<>());
            }
            
            if (statusData.containsKey("completedActions")) {
                status.setCompletedActions((List<String>) statusData.get("completedActions"));
            } else {
                status.setCompletedActions(new ArrayList<>());
            }
            
            if (statusData.containsKey("priority")) {
                status.setPriority((String) statusData.get("priority"));
            }
            
            if (statusData.containsKey("category")) {
                status.setCategory((String) statusData.get("category"));
            }
            
            if (statusData.containsKey("subCategory")) {
                status.setSubCategory((String) statusData.get("subCategory"));
            }
            
            if (statusData.containsKey("householdId")) {
                status.setHouseholdId((String) statusData.get("householdId"));
            }
            
            if (statusData.containsKey("relatedClientIds")) {
                status.setRelatedClientIds((List<String>) statusData.get("relatedClientIds"));
            }
            
            if (statusData.containsKey("beneficiaryIds")) {
                status.setBeneficiaryIds((List<String>) statusData.get("beneficiaryIds"));
            }
            
            if (statusData.containsKey("relationshipTypes")) {
                status.setRelationshipTypes((Map<String, String>) statusData.get("relationshipTypes"));
            }
            
            if (statusData.containsKey("tags")) {
                status.setTags((Map<String, String>) statusData.get("tags"));
            } else {
                status.setTags(new HashMap<>());
            }
            
            if (statusData.containsKey("metadata")) {
                status.setMetadata((Map<String, Object>) statusData.get("metadata"));
            }
            
            // Initialize history
            List<StatusHistoryItem> history = new ArrayList<>();
            StatusHistoryItem initialHistoryItem = new StatusHistoryItem();
            initialHistoryItem.setTimestamp(now.toString());
            initialHistoryItem.setChangedBy((String) statusData.get("createdBy"));
            initialHistoryItem.setNewStage(STAGE_INITIATED);
            initialHistoryItem.setChangeDescription("Status created");
            history.add(initialHistoryItem);
            status.setStatusHistory(history);
            
            // Initialize metadata
            Map<String, Object> metadata = new HashMap<>();
            
            // Perform sentiment analysis if summary is provided
            String summary = (String) statusData.get("statusSummary");
            if (summary != null && !summary.isEmpty()) {
                try {
                    if (isTestMode) {
                        // In test mode, use a fixed sentiment value
                        metadata.put("sentiment", "positive");
                    } else {
                        DetectSentimentRequest request = DetectSentimentRequest.builder()
                            .text(summary)
                            .languageCode("en")
                            .build();
                        
                        DetectSentimentResponse response = comprehendClient.detectSentiment(request);
                        metadata.put("sentiment", response.sentimentAsString().toLowerCase());
                    }
                } catch (Exception e) {
                    // Default to neutral if sentiment analysis fails
                    metadata.put("sentiment", SentimentType.NEUTRAL.toString().toLowerCase());
                }
            }
            
            status.setMetadata(metadata);
            
            // Store the status in DynamoDB
            dynamoDBService.putStatus(status);
            
            return status;
        } catch (Exception e) {
            throw new RuntimeException("Error creating status: " + e.getMessage(), e);
        }
    }
    
    /**
     * Retrieves a status by its ID.
     */
    public Status getStatus(String statusId) {
        try {
            return dynamoDBService.getStatus(statusId);
        } catch (Exception e) {
            throw new RuntimeException("Error retrieving status: " + e.getMessage(), e);
        }
    }
    
    /**
     * Retrieves a status by its source ID.
     * This allows linking back to external systems like workflow or order management systems.
     */
    public Status getStatusBySourceId(String sourceId) {
        if (sourceId == null || sourceId.isEmpty()) {
            return null;
        }
        
        try {
            return dynamoDBService.getStatusBySourceId(sourceId);
        } catch (Exception e) {
            throw new RuntimeException("Error retrieving status by source ID: " + e.getMessage(), e);
        }
    }
    
    /**
     * Retrieves a status by its tracking ID.
     * This allows customer service to easily look up statuses based on user-provided tracking IDs.
     */
    public Status getStatusByTrackingId(String trackingId) {
        if (trackingId == null || trackingId.isEmpty()) {
            return null;
        }
        
        try {
            return dynamoDBService.getStatusByTrackingId(trackingId);
        } catch (Exception e) {
            throw new RuntimeException("Error retrieving status by tracking ID: " + e.getMessage(), e);
        }
    }
    
    /**
     * Updates an existing status with the provided data.
     */
    public Status updateStatus(String statusId, Map<String, Object> statusData) {
        try {
            Status status = dynamoDBService.getStatus(statusId);
            
            if (status == null) {
                return null;
            }
            
            // Track previous stage for history
            String previousStage = status.getCurrentStage();
            
            // Update fields if provided
            if (statusData.containsKey("currentStage")) {
                status.setCurrentStage((String) statusData.get("currentStage"));
            }
            
            if (statusData.containsKey("statusSummary")) {
                status.setStatusSummary((String) statusData.get("statusSummary"));
                
                // Re-analyze sentiment if summary changed
                String sentiment = analyzeSentiment(status.getStatusSummary());
                status.addMetadata("sentiment", sentiment);
            }
            
            if (statusData.containsKey("statusDetails")) {
                status.setStatusDetails((Map<String, Object>) statusData.get("statusDetails"));
            }
            
            if (statusData.containsKey("requiredActions")) {
                status.setRequiredActions((List<String>) statusData.get("requiredActions"));
            }
            
            if (statusData.containsKey("completedActions")) {
                status.setCompletedActions((List<String>) statusData.get("completedActions"));
            }
            
            if (statusData.containsKey("priority")) {
                status.setPriority((String) statusData.get("priority"));
            }
            
            if (statusData.containsKey("estimatedCompletionDate")) {
                // Parse date from string if needed
                Object dateObj = statusData.get("estimatedCompletionDate");
                if (dateObj instanceof String) {
                    status.setEstimatedCompletionDate((String) dateObj);
                }
            }
            
            if (statusData.containsKey("actualCompletionDate")) {
                // Parse date from string if needed
                Object dateObj = statusData.get("actualCompletionDate");
                if (dateObj instanceof String) {
                    status.setActualCompletionDate((String) dateObj);
                }
            }
            
            if (statusData.containsKey("tags")) {
                status.setTags((Map<String, String>) statusData.get("tags"));
            }
            
            if (statusData.containsKey("metadata")) {
                Map<String, Object> newMetadata = (Map<String, Object>) statusData.get("metadata");
                Map<String, Object> existingMetadata = status.getMetadata();
                
                // Merge metadata
                if (existingMetadata == null) {
                    status.setMetadata(newMetadata);
                } else {
                    existingMetadata.putAll(newMetadata);
                }
            }
            
            // Update last updated info
            status.setLastUpdatedBy((String) statusData.get("updatedBy"));
            status.updateLastUpdatedDate();
            
            // Create history entry if stage changed
            if (!previousStage.equals(status.getCurrentStage())) {
                StatusHistoryItem historyItem = new StatusHistoryItem();
                historyItem.setTimestamp(Instant.now().toString());
                historyItem.setChangedBy(status.getLastUpdatedBy());
                historyItem.setPreviousStage(previousStage);
                historyItem.setNewStage(status.getCurrentStage());
                
                if (statusData.containsKey("changeReason")) {
                    historyItem.setChangeReason((String) statusData.get("changeReason"));
                }
                
                if (statusData.containsKey("changeDescription")) {
                    historyItem.setChangeDescription((String) statusData.get("changeDescription"));
                }
                
                // Add to history
                List<StatusHistoryItem> history = status.getStatusHistory();
                if (history == null) {
                    history = new ArrayList<>();
                    status.setStatusHistory(history);
                }
                history.add(historyItem);
            }
            
            // Store the updated status in DynamoDB
            dynamoDBService.putStatus(status);
            
            return status;
        } catch (Exception e) {
            throw new RuntimeException("Error updating status: " + e.getMessage(), e);
        }
    }
    
    /**
     * Retrieves all statuses for a specific client.
     */
    public List<Status> getClientStatuses(String clientId, String statusType, String fromDate, String toDate) {
        try {
            // Use the real implementation
            List<Status> statuses = dynamoDBService.getClientStatuses(clientId, statusType);
            
            // Apply date filtering if needed
            if (fromDate != null || toDate != null) {
                // Date filtering would be implemented here in a production environment
                // For now, we'll just return all statuses
            }
            
            return statuses;
        } catch (Exception e) {
            throw new RuntimeException("Error retrieving client statuses: " + e.getMessage(), e);
        }
    }
    
    /**
     * Retrieves all client statuses for a specific advisor.
     */
    public Map<String, List<Status>> getAdvisorClientStatuses(String advisorId, String statusType, String fromDate, String toDate) {
        try {
            // Use the real implementation
            Map<String, List<Status>> clientStatuses = dynamoDBService.getAdvisorClientStatuses(advisorId, statusType);
            
            // Apply date filtering if needed
            if (fromDate != null || toDate != null) {
                // Date filtering would be implemented here in a production environment
                // For now, we'll just return all statuses
            }
            
            return clientStatuses;
        } catch (Exception e) {
            throw new RuntimeException("Error retrieving advisor client statuses: " + e.getMessage(), e);
        }
    }
    
    /**
     * Searches for statuses based on various criteria.
     */
    public List<Status> searchStatuses(Map<String, Object> searchCriteria) {
        try {
            return dynamoDBService.searchStatuses(searchCriteria);
        } catch (Exception e) {
            throw new RuntimeException("Error searching statuses: " + e.getMessage(), e);
        }
    }
    
    /**
     * Analyzes the sentiment of the provided text using Amazon Comprehend.
     */
    private String analyzeSentiment(String text) {
        if (isTestMode) {
            return "positive";
        }
        
        try {
            DetectSentimentRequest request = DetectSentimentRequest.builder()
                .text(text)
                .languageCode("en")
                .build();
            
            DetectSentimentResponse response = comprehendClient.detectSentiment(request);
            return response.sentimentAsString().toLowerCase();
        } catch (Exception e) {
            // Log the error in production
            return "neutral";
        }
    }
}

File: ui/src/app/services/status-tracker.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { Status, Step } from '../models/status.model';
import { MockDataService } from './mock-data.service';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class StatusTrackerService {
  private apiUrl = environment.apiUrl;
  private useMockData = true; // Set to false to use real API

  constructor(
    private http: HttpClient,
    private mockDataService: MockDataService
  ) { }

  // Create a new status
  createStatus(statusData: any): Observable<Status> {
    if (this.useMockData) {
      // Return the first mock status as a placeholder
      return of(this.mockDataService.getMockStatuses()[0]);
    }
    return this.http.post<Status>(`${this.apiUrl}/status`, statusData);
  }

  // Get a status by ID
  getStatus(statusId: string): Observable<Status> {
    if (this.useMockData) {
      const status = this.mockDataService.getMockStatus(statusId);
      if (status) {
        return of(status);
      }
      // Return the first mock status if not found
      return of(this.mockDataService.getMockStatuses()[0]);
    }
    return this.http.get<Status>(`${this.apiUrl}/status/${statusId}`);
  }

  // Get a status by source ID
  getStatusBySourceId(sourceId: string): Observable<Status> {
    if (this.useMockData) {
      const statuses = this.mockDataService.getMockStatuses();
      const status = statuses.find(s => s.sourceId === sourceId);
      if (status) {
        return of(status);
      }
      // Return the first mock status if not found
      return of(statuses[0]);
    }
    const params = new HttpParams().set('sourceId', sourceId);
    return this.http.get<Status>(`${this.apiUrl}/status/source`, { params });
  }

  // Get a status by tracking ID
  getStatusByTrackingId(trackingId: string): Observable<Status> {
    if (this.useMockData) {
      const statuses = this.mockDataService.getMockStatuses();
      const status = statuses.find(s => s.trackingId === trackingId);
      if (status) {
        return of(status);
      }
      // Return the first mock status if not found
      return of(statuses[0]);
    }
    const params = new HttpParams().set('trackingId', trackingId);
    return this.http.get<Status>(`${this.apiUrl}/status/tracking`, { params });
  }

  // Update a status
  updateStatus(statusId: string, statusData: any): Observable<Status> {
    if (this.useMockData) {
      const status = this.mockDataService.getMockStatus(statusId);
      if (status) {
        return of({ ...status, ...statusData });
      }
      // Return the first mock status if not found
      return of(this.mockDataService.getMockStatuses()[0]);
    }
    return this.http.put<Status>(`${this.apiUrl}/status/${statusId}`, statusData);
  }

  // Get all statuses for a client
  getClientStatuses(clientId: string, statusType?: string, fromDate?: string, toDate?: string): Observable<Status[]> {
    if (this.useMockData) {
      const statuses = this.mockDataService.getMockStatuses();
      let filteredStatuses = statuses.filter(s => s.clientId === clientId);
      
      if (statusType) {
        filteredStatuses = filteredStatuses.filter(s => s.currentStage === statusType);
      }
      
      // Simple date filtering (not implementing for mock data)
      return of(filteredStatuses);
    }
    
    let params = new HttpParams();
    if (statusType) params = params.set('statusType', statusType);
    if (fromDate) params = params.set('fromDate', fromDate);
    if (toDate) params = params.set('toDate', toDate);
    
    return this.http.get<Status[]>(`${this.apiUrl}/client/${clientId}/statuses`, { params });
  }

  // Get all client statuses for an advisor
  getAdvisorClientStatuses(advisorId: string, statusType?: string, fromDate?: string, toDate?: string): Observable<Record<string, Status[]>> {
    if (this.useMockData) {
      const statuses = this.mockDataService.getMockStatuses();
      let filteredStatuses = statuses.filter(s => s.advisorId === advisorId);
      
      if (statusType) {
        filteredStatuses = filteredStatuses.filter(s => s.currentStage === statusType);
      }
      
      // Group by clientId
      const result: Record<string, Status[]> = {};
      filteredStatuses.forEach(status => {
        if (!result[status.clientId]) {
          result[status.clientId] = [];
        }
        result[status.clientId].push(status);
      });
      
      return of(result);
    }
    
    let params = new HttpParams();
    if (statusType) params = params.set('statusType', statusType);
    if (fromDate) params = params.set('fromDate', fromDate);
    if (toDate) params = params.set('toDate', toDate);
    
    return this.http.get<Record<string, Status[]>>(`${this.apiUrl}/advisor/${advisorId}/client-statuses`, { params });
  }

  // Search for statuses
  searchStatuses(searchCriteria: any): Observable<Status[]> {
    if (this.useMockData) {
      return of(this.mockDataService.searchMockStatuses(searchCriteria));
    }
    return this.http.post<Status[]>(`${this.apiUrl}/statuses/search`, searchCriteria);
  }

  // Step Management Methods
  
  // Add steps to a status
  addStepsToStatus(statusId: string, steps: Step[]): Observable<Status> {
    if (this.useMockData) {
      return this.mockDataService.addStepsToStatus(statusId, steps);
    }
    return this.http.post<Status>(`${this.apiUrl}/status/${statusId}/steps`, { steps });
  }
  
  // Update a step in a status
  updateStep(statusId: string, stepId: string, stepData: Partial<Step>): Observable<Status> {
    if (this.useMockData) {
      return this.mockDataService.updateStep(statusId, stepId, stepData);
    }
    return this.http.put<Status>(`${this.apiUrl}/status/${statusId}/steps/${stepId}`, stepData);
  }
  
  // Set current step
  setCurrentStep(statusId: string, stepId: string): Observable<Status> {
    if (this.useMockData) {
      return this.mockDataService.setCurrentStep(statusId, stepId);
    }
    return this.http.put<Status>(`${this.apiUrl}/status/${statusId}/current-step`, { stepId });
  }
  
  // Complete a step
  completeStep(statusId: string, stepId: string, notes?: string): Observable<Status> {
    if (this.useMockData) {
      return this.mockDataService.completeStep(statusId, stepId, notes);
    }
    return this.http.put<Status>(`${this.apiUrl}/status/${statusId}/steps/${stepId}/complete`, { notes });
  }
  
  // Skip a step
  skipStep(statusId: string, stepId: string, reason: string): Observable<Status> {
    if (this.useMockData) {
      return this.mockDataService.skipStep(statusId, stepId, reason);
    }
    return this.http.put<Status>(`${this.apiUrl}/status/${statusId}/steps/${stepId}/skip`, { reason });
  }
  
  // Block a step
  blockStep(statusId: string, stepId: string, reason: string): Observable<Status> {
    if (this.useMockData) {
      return this.mockDataService.blockStep(statusId, stepId, reason);
    }
    return this.http.put<Status>(`${this.apiUrl}/status/${statusId}/steps/${stepId}/block`, { reason });
  }
  
  // Reorder steps
  reorderSteps(statusId: string, stepOrder: {stepId: string, order: number}[]): Observable<Status> {
    if (this.useMockData) {
      return this.mockDataService.reorderSteps(statusId, stepOrder);
    }
    return this.http.put<Status>(`${this.apiUrl}/status/${statusId}/steps/reorder`, { stepOrder });
  }
} 